### PR TITLE
SelectForUpdate support

### PR DESCRIPTION
**Pull request by smootoo**
This implements (and tests) simple "select for update" support.
I decided not to try to implement any of the more advanced "for update of col1, col2" type syntax as the functionality is only provided on a few db backends. It can be added if needed in the future.
Although we haven't merged the CI for Oracle, DB2 and SQLServer yet, this has been tested on those backends as well.

### CODE REVIEW

**Comment by raimohanska** [comment]
Looks good to me!

**Comment by cvogt** [comment]
This is great :). LGTM. @szeiger should have a look at this.

**Comment by cvogt** [comment]
@szeiger *bump* :)

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/util/StandardTestDBs.scala [hunk 1]
```diff
@@ -292,6 +295,9 @@ object DerbyDB {
 abstract class HsqlDB(confName: String) extends InternalJdbcTestDB(confName) {
   val profile = HsqldbProfile
   val jdbcDriver = "org.hsqldb.jdbcDriver"
+  // Hsqldb has valid "select for update" syntax, but in testing, it either takes a whole table lock or no exclusive
+  // lock at all, so exclude from ForUpdate testing
+  override def capabilities = super.capabilities - TestDB.capabilities.selectForUpdateRowLocking
```
Comment (szeiger) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/util/StandardTestDBs.scala]: Should the profile declare selectForUpdateRowLocking at all if it doesn't do the expected thing? In these cases I think the limited support for a feature should be mentioned in the profile's scaladoc comment but the capability flag should not be set.

Comment (szeiger) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/util/StandardTestDBs.scala]: Oh, wait, this one is in TestDB.capabilities. So let me rephrase my last comment: Why is it separate from `rowLocking`? If HsqlDB doesn't properly lock in these cases, I think it shouldn't advertise `rowLocking` in the first place.

File: slick/src/main/scala/slick/ast/Comprehension.scala [hunk 2]
```diff
@@ -9,9 +9,11 @@ final case class Comprehension(sym: TermSymbol, from: Node, select: Node, where:
                                groupBy: Option[Node] = None, orderBy: ConstArray[(Node, Ordering)] = ConstArray.empty,
                                having: Option[Node] = None,
                                distinct: Option[Node] = None,
-                               fetch: Option[Node] = None, offset: Option[Node] = None) extends DefNode {
+                               fetch: Option[Node] = None,
+                               offset: Option[Node] = None,
+                               forUpdate: Option[Node] = None) extends DefNode {
```
Comment (szeiger) [inline slick/src/main/scala/slick/ast/Comprehension.scala]: Why is forUpdate not a Boolean? It doesn't look like any of the code generators uses the contents, nor does `mergeToComprehensions` ensure that it's kept consistent.

File: slick/src/main/scala/slick/lifted/Query.scala [hunk 3]
```diff
@@ -155,6 +155,11 @@ sealed abstract class Query[+E, U, C[_]] extends QueryBase[C[U]] { self =>
     new WrappingQuery[(G, Query[P, U, Seq]), (T, Query[P, U, Seq]), C](group, key.zip(value))
   }
 
+  /** Specify part of a select statement for update and marked for row level locking */
+  def forUpdate: Query[E, U, C] = {
```
Comment (szeiger) [inline slick/src/main/scala/slick/lifted/Query.scala]: I'm not convinced about the design. Is `forUpdate` really a query language feature? For instance, what happens when you put it into a subquery? Does this have sensible semantics that are different from putting it at the top level (or nowhere)? I would assume that it's rather a feature of query *execution*. A natural place for this would be an overload or flag in `Query.result` (e.g. `q.result(locked = true)`) or a new method (`q.lockedResult`) that is only available in `SqlProfile`. The flag could be passed directly to the code generator (similar to how forced insert vs soft insert is handled), thus avoiding the changes to the AST and `mergeToComprehensions`.

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/util/StandardTestDBs.scala (hunk 1 above)
Comment (smootoo) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/util/StandardTestDBs.scala]: I couldn't work out why the test wasn't passing for Hsqldb. It's not a backend I'm familiar with. It could just be a configuration problem in the test environment. The syntax generated is valid and as far as I can tell, it should work. I took the decision to add the capability to the profile, but exclude from testing in case it was just a test configuration problem. I can exclude from the Hsql profile if you think that is better.

File: slick/src/main/scala/slick/ast/Comprehension.scala (hunk 2 above)
Comment (smootoo) [inline slick/src/main/scala/slick/ast/Comprehension.scala]: This is to support the syntax which specifies explicit columns to lock to give finer grained locking control e.g. ```for update of col1, col2``` http://docs.oracle.com/cd/E11882_01/server.112/e41084/statements_10002.htm#SQLRF55372. As I mentioned in the text for the PR, I haven't implement this yet, but by implementing it as a node, gives us placeholder to add it.

File: slick/src/main/scala/slick/lifted/Query.scala (hunk 3 above)
Comment (smootoo) [inline slick/src/main/scala/slick/lifted/Query.scala]: Although not all the backends support forUpdate in complex queries, some do. For example, MYSQL does http://stackoverflow.com/questions/18318681/nested-update-with-select-deadlock/18646713#18646713. By having it as a query language feature, we can support those more complex queries as well as the forUpdate syntax that specifies explicit columns, as mentioned above.

Comment (szeiger) [inline slick/src/main/scala/slick/lifted/Query.scala]: Interesting. So it indeed makes sense to put it into the query language. It's a rather specific feature, with a SQL-specific name, but I can't think of a better design, either.

File: slick/src/main/scala/slick/ast/Comprehension.scala (hunk 2 above)
Comment (szeiger) [inline slick/src/main/scala/slick/ast/Comprehension.scala]: I'd use a Boolean for now until it's clear how to really handle the more complex case. It's probably similar to how `distinct` and `distinctOn` are implemented. I wouldn't be surprised if the most sensible definition of `forUpdate` also turns out to be `forUpdateOn(identity)`. This would cause a change in semantics from the current version (not a reason to hold back on this feature, just something to watch out for).

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/util/StandardTestDBs.scala (hunk 1 above)
Comment (szeiger) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/util/StandardTestDBs.scala]: If you think it *should* be OK for Hsqldb but get runtime failures, it's probably just a bug or limitation in Hsqldb. In that case I'd make the driver advertise the capability (as you did) and add a note to the scaladoc that there are limitations. Excluding the test with an extra capability is probably the best option. I don't have a better idea, either.

**Comment by smootoo** [comment]
forUpdate is now a Boolean not Option[Node] in Comprehension as requested by @szeiger

**Comment by szeiger** [comment]
LGTM

**Comment by rikiya-yamamoto** [comment]
PostgreSQL support,
Not yet?
